internal/sftp/client.go

```diff
@@ -24,9 +28,28 @@ func NewAuthError(e error) error {
 // authentication, and other intricacies associated with different SFTP
 // servers and protocols.
 type Client interface {
-	// Upload transfers data from the provided source reader to a specified
-	// destination on the SFTP server.
-	Upload(ctx context.Context, src io.Reader, dest string) (bytes int64, remotePath string, err error)
-	// Delete removes data from the provided dest on the SFTP server.
-	Delete(ctx context.Context, dest string) (err error)
+	// Delete removes dest from the SFTP server.
+	Delete(ctx context.Context, dest string) error
+	// Upload asynchronously copies data from the src reader to the specified
+	// dest on the SFTP server.
+	Upload(ctx context.Context, src io.Reader, dest string) (remotePath string, upload AsyncUpload, err error)
```

Comment:
I'm interested to see how our approach to this API evolves over time. Lately, I've noticed a trend in the community towards more synchronous APIs, and moving away from using channels due to their complexity (semantics) and potential for errors.

If we had to buidl a public API that prioritizes simplicity and ease of use, we could explore these two alternatives:

1. Provide a io.Writer, e.g. [gocloude.dev's bucket.NewWriter](https://gocloud.dev/howto/blob/#writing),
2. Provide a callback, e.g.:

```go
type UploadStatusCallback func(bytesWritten int64)
    
// Upload blocks.
Upload(ctx context.Context, src io.Reader, dest string, updater UploadStatusCallback) (remotePath string, err error)
```

Just to be clear, I'm not suggesting we need to change the current design.

Reply:
@sevein my *original* plan was to add an `UploadChunk()` method that would take start and width parameters, and then just loop it in the upload activity with a heartbeat + bytes upload count each iteration.   I thought it would fairly easy to do chunked read/writes with the `io` interfaces, but my reading on it gave me the impression that reading/writing chunks like that is not very efficient.  The best answer I found for how to do an (up/down)load progress update was to use `io.TeeWriter()` like I've used here.  Unfortunately I don't see any way to use the TeeWriter for updates without goroutines and channels for communication. :shrug: 

I'm not sure how you get around the blocking problem with `io.Writer` or a callback to provide heartbeats during the upload, without using goroutines somehow - it just seems like it would push that complexity to the caller instead of the Upload method.

Reply:
I've read somewhere that the Go SDK throttles heartbeats. I suspect that's why in the code examples they just show `RecordHeartbeat` being used in loops. The SDK makes sure we don't flood the Temporal frontend service with unnecessary heartbeats. I'll try to find where I saw that.

Reply:
Here: https://docs.temporal.io/activities#throttling.

Reply:
So activity code could look like:

```go
type progressReader struct {
    r        io.Reader
    callback func(int)
}

func (pr *progressReader) Read(p []byte) (int, error) {
    n, err := pr.r.Read(p)
    if err == nil {
        pr.callback(n)
    }
    return n, err
}

io.Copy(w, progressReader{myFileReader, myHeartbeater})
```

Not sure if that's any easier, maybe?